lib/learn_test/strategies/jest.rb

```diff
@@ -0,0 +1,60 @@
+module LearnTest
+  module Strategies
+    class Jest < LearnTest::Strategy
+      include LearnTest::JsStrategy
+
+      def service_endpoint
+        '/e/flatiron_jest'
+      end
+
+      def detect
+        return false if !js_package
+
+        has_js_dependency?(:jest)
+      end
+
+      def check_dependencies
+        Dependencies::NodeJS.new.execute
+      end
+
+      def run
+        run_jest
+      end
+
+      def cleanup
+        FileUtils.rm('.results.json') if File.exist?('.results.json')
+      end
+
+      def results
+        @results ||= {
+          username: username,
+          github_user_id: user_id,
+          learn_oauth_token: learn_oauth_token,
+          repo_name: runner.repo,
+          build: {
+            test_suite: [{
+              framework: 'jest',
+              formatted_output: output,
+              duration: output[:testResults][0][:endTime] - output[:startTime]
+            }]
+          },
+          examples: output[:numTotalTestSuites],
+          passing_count: output[:numPassedTests],
+          failure_count: output[:numFailedTests]
+        }
+      end
+
+      def output
+        @output ||= File.exist?('.results.json') ? Oj.load(File.read('.results.json'), symbol_keys: true) : nil
+      end
+
+      private
+
+      def run_jest
+        npm_install
+        system("node_modules/.bin/jest --json --outputFile=.results.json")
```

Comment:
What do y'all think about moving the `npm_install` step here (+ maybe in the other JS test strategies as well) into a `rescue` block that fires if the Jest call fails? I figure the amount of test runs that we need to `npm install` before is a pretty small chunk of the overall test runs, so we'd be optimizing for the "many" case. Something like:

```ruby
def run
  run_jest
rescue [specific error]
  npm_install
  run_jest
end

def run_jest
  `node_modules/.bin/jest --json --outputFile=.results.json`
end
```

There's probably a prettier way to do it, but that's the general idea

Reply:
Hey @gj , since `npm_install` checks for missing dependencies and doesn't do anything if it finds the `node_modules` folder, what do we gain from not running this every time?

Reply:
It's doing a fair amount of unnecessary IO — definitely not a huge deal, but I'm particularly sensitive to Learn gem performance b/c students allegedly cite it as a reason why they prefer running tests directly instead of using the `learn test` command.

If `npm_install` finds the `node_modules` dir, it doesn't just exit — it proceeds to iterate over every single dependency in `package.json` to check whether a corresponding directory exists in `node_modules`. If there are [27 dependencies in `package.json`](https://github.com/learn-co-curriculum/redux-thunk-lab/blob/c1bea2469258e083f3daca33c1160628a923cc59/package.json#L13-L43), that's 28 invocations of `File.exist?` that are gonna be unnecessary in the vast majority of cases.

All that said, it's almost definitely over-optimizing, so I'm happy to keep it as-is. Was just curious for your thoughts 😄

Reply:
Cool so I must have totally misread this part and I agree - one thing I'm wondering if we were to check for errors:

If all `jest` folders are missing from `node_modules` it will return `nil` upon running the Jest command. HOWEVER some of the other `jest` folders being missing (like `jest_config`) will trigger an actual error. My question - how likely is it that a random folder from the `jest` dependency will be missing? I imagine it's more likely that either `jest` is fully there or not, but if you think this is something we could run into I'll look at handling both failures.